llvm/test/CodeGen/AMDGPU/inline-constraints.ll

```diff
@@ -99,39 +99,39 @@ define i32 @inline_A_constraint_H1() {
 ; VI-LABEL: {{^}}inline_A_constraint_H2:
 ; VI: v_mov_b32 {{v[0-9]+}}, 0x3c00
 define i32 @inline_A_constraint_H2() {
-  %v0 = tail call i32 asm "v_mov_b32 $0, $1", "=v,A"(i16 bitcast (half 1.0 to i16))
+  %v0 = tail call i32 asm "v_mov_b32 $0, $1", "=v,A"(trunc i32 bitcast (float 1.0 to i32) to i16)
```

Comment:
TBH I don't think this test case is valid here. The instruction accepts a `b32` operand, and we pass a `i16`, which is bitcasted from `fp16`. We "expect" the backend to take it as an inline literal, but apparently it has already been casted to `i16`, the backend can only see `0x3C00` here, how does it know whether it is `1.0`? The spec says, `1.0` needs to be taken as inline literal, but it does't state in what encoding. My understanding is, it should be based on the type of the operand. If the operand expects a `bf16`, then `0x3F80` is `1.0`. If the operand expects a `fp16`, then `0x3C00` should be it.
The problem here is, if we don't treat it as an inline literal, the instruction will not be emitted.

Reply:
@rampitec WDYT?

Reply:
I do not see a problem with the original test. It does not expect an inline literal, it uses 0x3c00 which is a correct value not interpreted in any way other than literal. I.e. `i16 bitcast (half 1.0 to i16)` to me looks exactly as 0x3c00, which is the substituted and used in the mov as a constant. If it were expected inline constant here the check would be (incorrectly) `v_mov_b32 {{v[0-9]+}}, 1.0`.

Reply:
And then ultimately you cannot reason what encoding is used without --show-mc-encoding.

Reply:
But the thing is, we have an inline asm constraint `A` here, which indeed requires the operand should be an inline literal. Now we only know the type of the operand is `i16`, but for floating-point inline literal, it can be either FP16 or BF16. Like you said, what we actually see is `0x3C00`, but w/o enough information, we can't know whether it is the `half 1.0` or something else.

Reply:
We probably should just complain on 'A' here because we cannot disambiguate it...

Reply:
But wait... Where is it written, it shall be an INLINE literal? Literal maybe, but inline... We do not have such controls in the first place.

Reply:
It's in https://llvm.org/docs/LangRef.html#supported-constraint-code-list:
> A: An integer or a floating-point inline constant.

I think we might want to remove those test cases or just leave one to check we actually don't take them as inline literal. WDYT? I can do it along with this patch.

Reply:
Yes, I think the inline literal only makes sense when it is along with corresponding type. The underlying issue within this test case (as well as the implementation of checking 16-bit inline literal) is, it tries to recognize some special combination of bits such that as long as it is, say `0x3C00` and it is 16-bit operand, we take it as inline literal ` half 1.0`. This is problematic.

One of the goal of this patch is to do this right: check the inline literal with corresponding type, instead of just checking whether this 16-bit thing is an inline literal or not.

Reply:
If you drop 'A' from the equation things get easy. This is 0x3c00 regardless of the instruction context. When you finally assemble it you see this cannot be inline and just use literal. But that is how our asm and inline asm work in general. `v_mov_b32 {{v[0-9]+}}, 0x3c00` is a right thing to me... if we just drop 'A'. And this right thing also tells nothing about the encoding too, the test just does not check the encoding. Check this:

```
llvm-mc -arch=amdgcn -mcpu=gfx1100 -show-encoding <<< 'v_add_f16 v0, v1, 0x3c00'
        v_add_f16_e64 v0, v1, 1.0               ; encoding: [0x00,0x00,0x32,0xd5,0x01,0xe5,0x01,0x00]
```
I have used 0x3c00, but the inline 1.0 was encoded as a shortest form. Regardless of how I wrote it in the asm.

```
llvm-mc -arch=amdgcn -mcpu=gfx1100 -show-encoding <<< 'v_add_f16 v0, v1, 0x3c01'
        v_add_f16_e64 v0, v1, 0x3c01            ; encoding: [0x00,0x00,0x32,0xd5,0x01,0xff,0x01,0x00,0x01,0x3c,0x00,0x00]
```
This does not fit and it is literal.

So when we are handling inline asm we just do not deal with the encoding at all, and printing 0x3c00 does not tell anything about the encoding, nor it tests it.

IMHO the proper fix shall drop the word 'inline' from the documentation.

Reply:
In the test case, `0x3c00` is not inlined *in the encoding* because the operand is `i32` instead of `i16`. However, when we parse the inline asm, the parser can only see `i16` because that's what we tell the parser via the type. Then we have a constraint `A`, so the parser is trying to parse it as a 16-bit inline literal. As for whether eventually it is the inline literal in the encoding, it will only happen after we finish parsing.

`v_mov_b32 {{v[0-9]+}}, 0x3c00` is right but it should not be the intention of this test case because eventually it is not an inline instant. It just "tricks" the inline asm parser to falsely take it as inline literal.

Reply:
Yes, it seems broken, and it is already broken.